cmd/mt-kafka-mdm-report-out-of-order/inputooofinder.go

```diff
@@ -0,0 +1,193 @@
+package main
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/grafana/metrictank/schema"
+	"github.com/grafana/metrictank/schema/msg"
+	log "github.com/sirupsen/logrus"
+)
+
+type Tracker struct {
+	Head      Msg    // last successfully added message
+	Bad       Msg    // current (last seen) point that could not be added (assuming no re-order buffer)
+	NumBad    int    // number of failed points since last successful add
+	DeltaTime uint32 // delta between Head and Bad time properties in seconds (point timestamps)
+	DeltaSeen uint32 // delta between Head and Bad seen time in seconds (consumed from kafka)
+}
+
+type Msg struct {
+	Part int32
+	Seen time.Time
+	Md   schema.MetricData // either this one or below will be valid depending on input
+	Mp   schema.MetricPoint
+}
+
+func (m Msg) Time() uint32 {
+	if m.Md.Id != "" {
+		return uint32(m.Md.Time)
+	}
+	return m.Mp.Time
+}
+
+// find out of order metrics
+type inputOOOFinder struct {
+	graceDuration time.Duration
+	prefix        string
+	substr        string
+	doUnknownMP   bool
+
+	data        map[schema.MKey]Tracker
+	definitions map[schema.MKey]schema.MetricDefinition
+
+	groupByTag    string
+	groupedByName *map[string]int
+	groupedByTag  *map[string]int
+
+	lock sync.Mutex
+}
+
+func newInputOOOFinder(graceDuration time.Duration, prefix string, substr string, doUnknownMP bool, definitions map[schema.MKey]schema.MetricDefinition, groupByTag string, groupedByName *map[string]int, groupedByTag *map[string]int) *inputOOOFinder {
+	return &inputOOOFinder{
+		graceDuration,
+		prefix,
+		substr,
+		doUnknownMP,
+
+		make(map[schema.MKey]Tracker),
+		definitions,
+
+		groupByTag,
+		groupedByName,
+		groupedByTag,
+
+		sync.Mutex{},
+	}
+}
+
+func (ip *inputOOOFinder) ProcessMetricData(metric *schema.MetricData, partition int32) {
+	if ip.prefix != "" && !strings.HasPrefix(metric.Name, ip.prefix) {
+		return
+	}
+	if ip.substr != "" && !strings.Contains(metric.Name, ip.substr) {
+		return
+	}
+
+	mkey, err := schema.MKeyFromString(metric.Id)
+	if err != nil {
+		log.Errorf("could not parse id %q: %s", metric.Id, err.Error())
+		return
+	}
+
+	ip.lock.Lock()
+	defer ip.lock.Unlock()
+
+	// update index
+	_, ok := ip.definitions[mkey]
+	if !ok {
+		ip.definitions[mkey] = schema.MetricDefinition{
+			Name: metric.Name,
+			Tags: metric.Tags,
+		}
+	}
+
+	now := Msg{
+		Part: partition,
+		Seen: time.Now(),
+		Md:   *metric,
+	}
+	tracker, ok := ip.data[mkey]
+	if !ok {
+		ip.data[mkey] = Tracker{
+			Head: now,
+		}
+	} else {
+		if uint32(metric.Time)+uint32(ip.graceDuration.Seconds()) > tracker.Head.Time() {
+			tracker.Head = now
```

Comment:
Should we only update `Head` if `metric.Time > Head.Time()`? It seems like this might cause `Head` to move backwards if slightly out of order data (within graceDuration) is found.